### PR TITLE
Add check points in tpm_verify_device test case

### PR DESCRIPTION
**Pull request by qcheng-redhat**
Add two check points:
1. Check windows event log
2. Execute power shell command: tpmsupportedfeature

ID: 2676

### CODE REVIEW

**Comment by qcheng-redhat** [comment]
Test results:
 (01/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.q35: STARTED
 (01/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.q35: PASS (145.96 s)
 (02/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_check_ovmf.q35: STARTED
 (02/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_check_ovmf.q35: PASS (148.02 s)
 (03/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_reboot.q35: STARTED
 (03/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_reboot.q35: PASS (283.31 s)
 (04/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_multi_vms.q35: STARTED
 (04/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_multi_vms.q35: FAIL: No Found pattern "No matches found" from "\n   Index Time          EntryType   Source                 InstanceID Message                                           \n   ----- ----          ---------   ------                 ---------- -------                ... (251.85 s)
 (05/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.q35: STARTED
 (05/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.q35: PASS (155.44 s)
 (06/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_check_ovmf.q35: STARTED
 (06/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_check_ovmf.q35: PASS (148.26 s)
 (07/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_reboot.q35: STARTED
 (07/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_reboot.q35: PASS (240.90 s)
 (08/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_multi_vms.q35: STARTED
 (08/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_multi_vms.q35: PASS (288.10 s)
 (09/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.tpm_check_buffer_size.q35: STARTED
 (09/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.tpm_check_buffer_size.q35: PASS (8.29 s)
 (10/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.tpm_check_buffer_size.q35: STARTED
 (10/10) Host_RHEL.m9.u5.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.tpm_check_buffer_size.q35: PASS (8.36 s)

Please ignore the failed test, it's due to a product bug.

**Comment by qcheng-redhat** [comment]
Hi @leidwang, please help review this patch when you have time. Thanks!

**Review by leidwang — APPROVED** [review]
LGTM.

**Comment by qcheng-redhat** [comment]
Hi @vivianQizhu, can you help review this patch when you have time ? Thanks!

File: qemu/tests/cfg/tpm_verify_device.cfg [hunk 1]
```diff
@@ -30,6 +30,10 @@
         pattern_output_wmic = IsActivated_InitialValue=TRUE;IsEnabled_InitialValue=TRUE
         cmd_get_tpm = 'powershell -command "get-tpm"'
         pattern_output_get_tpm = TpmPresent\s+:\s+True;TpmReady\s+:\s+True
+        cmd_check_event_log = 'powershell -command "Get-EventLog -LogName System -EntryType Error -Source TPM"'
+        pattern_output_event_log = 'No matches found'
```
Comment (vivianQizhu) [inline qemu/tests/cfg/tpm_verify_device.cfg:34]: What if the pattern was changed? This is controlled by third party. Do we have a more accurate way to check if there is any failure, like the error type or return value?

File: qemu/tests/tpm_verify_device.py [hunk 2]
```diff
@@ -61,12 +71,26 @@ def search_keywords(patterns, string, flags=re.M, split_string=';'):
...
                 error_context.context("Reboot guest '%s'." % vm.name, test.log.info)
                 vm.reboot(session, reboot_method).close()
                 continue
 
+            error_context.context("%s: Check Windows event log inside guest after test." % vm.name,
+                                  test.log.info)
+            if cmd_check_event_log:
+                event_log = session.cmd_output(cmd_check_event_log)
+                test.log.info(event_log)
+                search_keywords(event_log_pattern, event_log)
```
Comment (vivianQizhu) [inline qemu/tests/tpm_verify_device.py:92]: This is redundant for me. You have 3 similar implementation, we need a function for it then, e.g.
```
def check_error_msg(chk_cmd, chk_pattern):
    if not chk_cmd or not chk_pattern:
        return
    full_msg = session.cmd_output(chk_cmd)
    test.log.info(full_msg)
    search_keywords(chk_pattern, full_msg)
```
```
check_error_msg(cmd_check_event_log, event_log_pattern)
```
```
check_error_msg(cmd_get_tpmsupportedfeature, tpmspportedfeature)
```

**Review by vivianQizhu — CHANGES_REQUESTED** [review]

File: qemu/tests/tpm_verify_device.py (hunk 2 above)
Comment (qcheng-redhat) [reply qemu/tests/tpm_verify_device.py]: If I understand this comment correctly, search_keywords() is the function. And it is called at L64, L79 and L92 for different patterns. Or maybe I did not understand the point? 

I removed some duplicated printing log infos now.

File: qemu/tests/cfg/tpm_verify_device.cfg (hunk 1 above)
Comment (qcheng-redhat) [reply qemu/tests/cfg/tpm_verify_device.cfg:34]: We can define different Entrytypes  or other parameters in our command, it depends on the requirements, or what kind of message we are interested in.  If there are some records found by the powershell command, it means there are some error message in eventlog for vtpm device. Return value of the command depends on the  expected output. For now, if no product bug, this command returns 1.

**Comment by qcheng-redhat** [comment]
Hi @vivianQizhu, could you please help review the patch again when you have time? Thanks!

**Review by vivianQizhu — APPROVED** [review]

**Pull request merged**